fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSession.java

```diff
@@ -61,25 +84,36 @@ public DefaultOCFLObjectSession(final String objectIdentifier, final Path stagin
         this.objectIdentifier = objectIdentifier;
         this.stagingPath = stagingPath;
         this.ocflRepository = ocflRepository;
-    }
-
-    private Set<String> getDeletePaths() {
-        if (deletePaths == null) {
-            // TODO check to see if a delete log exists, in case resuming after reboot
-            deletePaths = new HashSet<>();
-        }
-        return deletePaths;
+        this.deletePaths = new HashSet<>();
+        this.objectDeleted = false;
+        this.sessionClosed = false;
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public void write(final String subpath, final InputStream stream) {
-        // TODO write contents to the subpath within the staging path
-
+    public synchronized void write(final String subpath, final InputStream stream) throws PersistentStorageException {
         // Determine the staging path for the incoming content
-        resolveStagedPath(subpath);
+        final var stagedPath = resolveStagedPath(subpath);
+        try {
+            assertSessionOpen();
+
+            final var parentPath = stagedPath.getParent();
+
+            // Fill in any missing parent directories
+            Files.createDirectories(parentPath);
+            // write contents to subpath within the staging path
+            Files.copy(stream, stagedPath, StandardCopyOption.REPLACE_EXISTING);
+        } catch (final IOException e) {
+            throw new PersistentStorageException("Unable to persist content to " + stagedPath, e);
+        } finally {
+            try {
+                stream.close();
+            } catch (final IOException e) {
+                log.error("Failed to close inputstream while writing {}", subpath, e);
```

Comment:
Would this not swallow (while logging) the error and possibly leave the stream open? Could that end up being a memory leak?

Reply:
Also noting my lack of experience tracing memory leaks in the first place.

Reply:
that could be the result, but we also can't close the stream, so I think we're kind of stuck with it.